avm/src/daemon/daemon.rs

```diff
@@ -116,118 +142,150 @@ async fn post_v1_stats(
-          Err(err) => {
-            error!(ErrorType::NoDnsVms, "{}", err).await;
-            Vec::new()
-          }
-        };
-        // triggered the first time since cluster_size == 0
-        // and every time the cluster changes size
-        if vms.len() != cluster_size {
-          cluster_size = vms.len();
-          let ips = vms
-            .iter()
-            .map(|vm| vm.private_ip_addr.to_string())
-            .collect();
-          control_port.update_ips(ips);
-          leader_ip = control_port.get_leader().to_string();
+pub async fn start() {
+  let mut control_port = ControlPort::start().await;
+  let self_ip = get_private_ip().await;
+  match get_private_ip().await {
+    Ok(self_ip) => {
+      let mut daemon_props: Option<&DaemonProperties> = None;
+      let duration = Duration::from_secs(10);
+      let mut counter: u8 = 0;
+      // Check every 10s over 5 min if props are ready
+      while counter < 30 && daemon_props.is_none() {
+        if let Some(props) = DAEMON_PROPS.get() {
+          daemon_props = Some(props);
         }
-        if leader_ip == self_ip.to_string() {
-          match get_v1_stats().await {
-            Ok(s) => stats.push(s),
-            Err(err) => error!(ErrorType::NoStats, "{}", err).await,
-          };
-        }
-        if stats.len() >= 4 {
-          let mut factor = String::from("1");
-          let stats_factor = post_v1_stats(stats.to_owned(), &cluster_id, &deploy_token).await;
-          stats = Vec::new();
-          if let Ok(stats_factor) = stats_factor {
-            factor = stats_factor;
-          } else if let Err(err) = stats_factor {
-            error!(ErrorType::PostFailed, "{}", err).await;
-          }
-          println!(
-            "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
-            cluster_id,
-            vms.len(),
-            factor
-          );
-          if factor != "1" {
-            post_v1_scale(&cluster_id, &agzb64, &deploy_token, &factor).await;
+        counter += 1;
+        sleep(duration).await;
+      }
+      if let Some(daemon_props) = daemon_props {
+        let cluster_id = &daemon_props.clusterId;
+        CLUSTER_ID.set(String::from(cluster_id)).unwrap();
+        let cluster_id = &daemon_props.clusterId;
+        let domain = &daemon_props.domain;
+        let deploy_token = &daemon_props.deployToken;
+        let agz_b64 = &daemon_props.agzB64;
+        task::spawn(async move {
+          let period = Duration::from_secs(60);
+          let mut stats = Vec::new();
+          let mut cluster_size = 0;
+          let mut leader_ip = String::new();
+          let dns = DNS::new(&domain);
+          if let Ok(dns) = &dns {
+            loop {
+              let vms = match dns.get_vms(&cluster_id).await {
+                Ok(vms) => vms,
+                Err(err) => {
+                  error!(NoDnsVms, "{}", err).await;
+                  Vec::new()
+                }
+              };
```

Comment:
nit: can be `unwrap_or_else`

Reply:
But, how could I handle the `error!` since it's a future? I cannot define an `async` block inside a `unwrap_or_else`, right?